lightgbm/src/main/scala/com/microsoft/ml/spark/lightgbm/PartitionProcessor.scala

```diff
@@ -8,57 +8,124 @@ import com.microsoft.ml.spark.lightgbm.TrainUtils.{afterGenerateTrainDataset, af
+      if (!isMainWorker) {
+        sharedState.incrementDoneSignal(log)
+      }
+      isMainWorker
+    } else {
+      true
+    }
+  }
+
+  def prepareDatasets(inputRows: Iterator[Row],
+                      validationData: Option[Broadcast[Array[Row]]],
+                      columnParams: ColumnParams,
+                      schema: StructType,
+                      trainParams: TrainParams,
+                      sharedState: SharedState): (AggregatedColumns, Option[AggregatedColumns]) = {
+    val aggregatedColumns = {
+      val prepAggregatedColumns = sharedState.datasetAggregator.get.prep(inputRows)
+      sharedState.datasetAggregator.get.merge(Array(prepAggregatedColumns))
+    }
+
+    val aggregatedValidationColumns = validationData.map { data =>
+      val validationDatasetAggregator = new DatasetAggregator(columnParams, schema, false,
+        trainParams.executionParams.chunkSize, trainParams.executionParams.matrixType, sharedState)
+      val prepAggregatedColumns = validationDatasetAggregator.prep(data.value.toIterator)
+      validationDatasetAggregator.merge(Array(prepAggregatedColumns))
+    }
+    (aggregatedColumns, aggregatedValidationColumns)
+  }
+
   def trainLightGBM(batchIndex: Int, networkParams: NetworkParams, columnParams: ColumnParams,
                     validationData: Option[Broadcast[Array[Row]]], log: Logger,
-                    trainParams: TrainParams, numTasksPerExec: Int, schema: StructType)
+                    trainParams: TrainParams, numTasksPerExec: Int, schema: StructType,
+                    sharedState: SharedState)
                    (inputRows: Iterator[Row]): Iterator[LightGBMBooster] = {
-    val emptyPartition = !inputRows.hasNext
-    val isEnabledWorker = !emptyPartition
-    // Initialize the native library
-    LightGBMUtils.initializeNativeLibrary()
-    // Initialize the network communication
-    val (nodes, localListenPort) = getNetworkInfo(networkParams, numTasksPerExec, log, isEnabledWorker)
-    if (emptyPartition) {
-      log.warn("LightGBM task encountered empty partition, for best performance ensure no partitions empty")
-      List[LightGBMBooster]().toIterator
-    } else {
-      log.info(s"LightGBM task listening on: $localListenPort")
-      // Return booster only from main worker to reduce network communication overhead
-      val returnBooster = getReturnBooster(isEnabledWorker, nodes, log, numTasksPerExec, localListenPort)
-      try {
-        // If worker enabled, initialize the network ring of communication
-        networkInit(nodes, localListenPort, log, LightGBMConstants.NetworkRetries, LightGBMConstants.InitialDelay)
-        translate(batchIndex, columnParams, validationData, log, trainParams, returnBooster, schema, inputRows)
-      } finally {
-        // Finalize network when done
-        if (isEnabledWorker) LightGBMUtils.validate(lightgbmlib.LGBM_NetworkFree(), "Finalize network")
+    var isEnabledWorker = false
```

Comment:
does this need to be a var or could it be a val on line 73

Reply:
I think it needs to be a var so I can pass it to the finally block:
```
resetState(useSingleDatasetMode, isEnabledWorker, log, sharedState)
```

Reply:
but this could be moved up a few lines to be outside the try. Also consider the StreamUtils.using trick to get rid of this and other vals that only exist because of a finally

Reply:
the resetState is no longer needed so I can make it a val now.
I needed to put the other assignment (isWorkerEnabled method) inside the try because it includes state updates, like "val mainExecutorWorker = sharedState.mainExecutorWorker", "sharedState.incrementArrayProcessedSignal(log)" and "sharedState.incrementDoneSignal(log)"